File: nodejs/src/strategy/CreditCardPayment.ts
import { PaymentStrategy, PaymentResult } from "./PaymentStrategy";

class CreditCardPayment implements PaymentStrategy {
    private cardNumber: string;
    private feeRate: number = 0.02;

    constructor(cardNumber: string) {
        this.cardNumber = cardNumber;
    }

    getName(): string { return "Thẻ tín dụng"; }

    pay(amount: number): PaymentResult {
        if (!this.cardNumber || this.cardNumber.length !== 16) {
            return { success: false, message: "Số thẻ không hợp lệ" };
        }
        const fee = Math.round(amount * this.feeRate);
        return {
            success: true,
            message: `Thanh toán ${amount.toLocaleString()} VND. Phí: ${fee.toLocaleString()} VND. Tổng: ${(amount + fee).toLocaleString()} VND`,
        };
    }
}

export default CreditCardPayment;

File: spring/demo/src/main/java/com/example/demo/strategy/CreditCardPayment.java
package com.example.demo.strategy;

public class CreditCardPayment implements PaymentStrategy {
    private final String cardNumber;
    private final double feeRate = 0.02;

    public CreditCardPayment(String cardNumber) {
        this.cardNumber = cardNumber;
    }

    @Override
    public String getName() {
        return "Thẻ tín dụng";
    }

    @Override
    public PaymentResult pay(int amount) {
        if (cardNumber == null || cardNumber.length() != 16) {
            return new PaymentResult(false, "Số thẻ không hợp lệ");
        }

        int fee = (int) Math.round(amount * feeRate);
        return new PaymentResult(true, String.format("Thanh toán %,d VND. Phí: %,d VND. Tổng: %,d VND", amount, fee, amount + fee));
    }
}
